op-batcher/channel_manager.go

```diff
@@ -35,90 +69,174 @@ func (s *channelManager) Clear() {
-	// Cap length at 100 blocks
-	l := len(s.blocks)
-	if l > 100 {
-		l = 100
-	}
-	for ; i < l; i++ {
-		if err := ch.AddBlock(s.blocks[i]); err == derive.ErrTooManyRLPBytes {
+	for ; i < len(blocks); i++ {
+		if err := ch.AddBlock(blocks[i]); err == derive.ErrTooManyRLPBytes {
 			break
 		} else if err != nil {
-			return nil, txID{}, err
+			return derive.ChannelID{}, nil, nil, err
 		}
-		// TODO: limit the RLP size of the channel to be lower than the limit to enable
-		// channels to be fully submitted on time.
 	}
 	if err := ch.Close(); err != nil {
-		return nil, txID{}, err
+		return derive.ChannelID{}, nil, nil, err
 	}
 
-	var t []taggedData
-	frameNumber := uint16(0)
+	var frames [][]byte
 	for {
 		var buf bytes.Buffer
 		buf.WriteByte(derive.DerivationVersion0)
-		err := ch.OutputFrame(&buf, 120_000)
+		err := ch.OutputFrame(&buf, maxFrameSize-1)
 		if err != io.EOF && err != nil {
-			return nil, txID{}, err
+			return derive.ChannelID{}, nil, nil, err
 		}
-
-		t = append(t, taggedData{
-			data: buf.Bytes(),
-			id:   txID{ch.ID(), frameNumber},
-		})
-		frameNumber += 1
+		frames = append(frames, buf.Bytes())
 		if err == io.EOF {
 			break
 		}
 	}
+	return ch.ID(), frames, blocks[i:], nil
+}
 
-	s.datas = append(s.datas, t...)
-	// Say i = 0, 1 are added to the channel, but i = 2 returns ErrTooManyRLPBytes. i remains 2 & is inclusive, so this works.
-	// Say all blocks are added, i will be len(blocks) after exiting the loop (but never inside the loop).
-	s.blocks = s.blocks[i:]
-
-	if len(s.datas) == 0 {
+// nextTxData pops off s.datas & handles updating the internal state
+func (s *channelManager) nextTxData() ([]byte, txID, error) {
+	if len(s.datas) != 0 {
+		r := s.datas[0]
+		s.pendingTransactions[r.id] = r.data
+		s.datas = s.datas[1:]
+		return r.data, r.id, nil
+	} else {
 		return nil, txID{}, io.EOF // TODO: not enough data error instead
 	}
+}
+
+// TxData returns the next tx.data that should be submitted to L1.
+// It is very simple & currently ignores the l1Head provided (this will change).
+// It may buffer very large channels as well.
+func (s *channelManager) TxData(l1Head eth.L1BlockRef) ([]byte, txID, error) {
+	s.log.Trace("Requested tx data")
+
+	// Short circuit if there is a pending channel.
+	// We either submit the next frame from that channel or
+	if s.channelPending {
+		return s.nextTxData()
+	}
+	// If we have no saved blocks, we will not be able to create valid frames
+	if len(s.blocks) == 0 {
+		return nil, txID{}, io.EOF
+	}
+
+	// Select range of blocks
+	end := len(s.blocks)
+	if end > 100 {
+		end = 100
+	}
+	blocks := s.blocks[:end]
+	s.blocks = s.blocks[end:]
+
+	chID, frames, leftOverBlocks, err := blocksToFrames(blocks, 120_000)
+	// If the range of blocks serialized to be too large, restore
+	// blocks that could not be included inside the channel
+	if len(leftOverBlocks) != 0 {
+		s.blocks = append(leftOverBlocks, s.blocks...)
+	}
+	if err != nil {
+		return nil, txID{}, err
+	}
+
```

Comment:
Perhaps sanity check on the length of frames being less than `math.MaxUint16` to prevent an overflow with the typecast, it should never happen but it would be a nice sanity check

Reply:
Gonna leave a TODO comment about this. It's a good idea, but has some trickiness when implementing it.